File: CV1/Python/CV1_Uloha2.py
import numpy as np
import matplotlib.pyplot as plt

def testfn3c(Pop):
    """
    MATLAB/Python interpretácia:
    Výpočet fitness pre ľubovoľný rozmer vstupu Pop.
    SOURCE: https://github.com/STU-FEI-OUI/UMINT-GA/blob/main/Kod/testfn3c.py
    """
    Pop = np.asarray(Pop, dtype=float)

    if Pop.ndim == 1:
        lpop = 1
        lstring = Pop.shape[0]
        X = Pop.reshape(1, lstring)
    else:
        lpop = Pop.shape[0]
        lstring = int(np.prod(Pop.shape[1:]))
        X = Pop.reshape(lpop, lstring)

    x0 = 30.0
    y0 = 100.0
    d_val = X - x0
    # Vektorizovaný výpočet sumy pre každého jedinca (riadok)
    Fit = np.sum(-(d_val) * np.sin(np.sqrt(np.abs(d_val))) + y0, axis=1)

    return Fit

# --- Nastavenia ---
x_min, x_max = -1000, 1000
max_iter = 30
d_step = 10

# 1. Príprava prostredia a vizualizácia pozadia
plt.ion() # Interaktívny režim pre animáciu
fig, ax = plt.subplots(figsize=(12, 7))

x_plot = np.linspace(x_min, x_max, 1000)
# Tu posielame celé pole naraz (2D tvar), vráti pole výsledkov
y_plot = testfn3c(x_plot.reshape(-1, 1)) 

ax.plot(x_plot, y_plot, 'r', linewidth=1, alpha=0.4, label='F(x)')
ax.grid(True, linestyle=':', alpha=0.6)
ax.set_title(f'Multi-start Hill Climbing (N={max_iter}) - testfn3c')

best_min_global = np.inf
x_min_global = None

# 2. Multi-start cyklus
for i in range(1, max_iter + 1):
    x_curr_val = np.random.uniform(x_min, x_max)
    
    # OPRAVA: Posielame [x_curr_val], aby to bolo 1D pole a funkcia nepadla
    f_curr = testfn3c([x_curr_val])[0] 
    
    # Štartovací bod (čierny krúžok)
    ax.plot(x_curr_val, f_curr, 'ko', markersize=5)
    
    while True:
        x_left = x_curr_val - d_step
        x_right = x_curr_val + d_step

        # Hranice a výpočet fitness pre susedov (opäť balíme do listu [])
        f_l = testfn3c([x_left])[0] if x_left >= x_min else np.inf
        f_r = testfn3c([x_right])[0] if x_right <= x_max else np.inf

        vals = [f_curr, f_r, f_l]
        idx = np.argmin(vals)

        if idx == 1: # Smer doprava
            x_curr_val, f_curr = x_right, f_r
        elif idx == 2: # Smer doľava
            x_curr_val, f_curr = x_left, f_l
        else:
            # Lokálne minimum nájdené (zelený bod)
            ax.plot(x_curr_val, f_curr, 'go', markersize=4)
            break
        
        # Vizuálna stopa hľadania
        ax.plot(x_curr_val, f_curr, 'b.', markersize=2, alpha=0.3)
        plt.pause(0.005)

    if f_curr < best_min_global:
        best_min_global = f_curr
        x_min_global = x_curr_val
    
    print(f"Iterácia {i:2d}: Koniec v x = {x_curr_val:8.2f}")

# 3. Finálne vyhodnotenie
print("\n" + "="*45)
print(f"GLOBÁLNE MINIMUM: x = {x_min_global:.4f}, F(x) = {best_min_global:.4f}")
print("="*45)

ax.plot(x_min_global, best_min_global, 'rs', markersize=12, fillstyle='none', 
        markeredgewidth=2, label='Global Best')
ax.legend()
plt.ioff() 
plt.show()

File: CV1/Matlab/CV1_Uloha2.m
clc; clear; close all;

% --- Nastavenia ---
x_range = -1000:1:1000;
max_iter = 30; % Počet náhodných spustení
d = 5;         % Krok hľadania

% 1. Vykreslenie základnej funkcie
figure;
y_plot = arrayfun(@(v) testfn3c(v), x_range);
plot(x_range, y_plot, 'r', 'LineWidth', 1);
hold on; grid on;

bestmin = inf;
xmin = inf;

% 2. Multi-start cyklus
for i = 1:max_iter
    % Náhodný štartovací bod (čierny bod)
    x_current = randi([-1000, 1000]);
    F_current = testfn3c(x_current);
    
    % Vykreslenie štartu iterácie
    plot(x_current, F_current, 'bo', 'MarkerFaceColor', 'k');
    
    % Volanie hľadania (tvoja logika findmin)
    [x_min_found, F_min_found] = hill_climbing_search(x_current, F_current, d);
    fprintf("Iterácia %d ukončená v minime: x=%.2f, F(x)=%.4f\n", i, x_min_found, F_min_found);
    % Kontrola, či je toto minimum lepšie ako doterajšie najlepšie
    if F_min_found < bestmin
        bestmin = F_min_found;
        xmin = x_min_found;
    end
end

% 3. Záverečný výpis
fprintf("\n==============================================");
fprintf("\nNajmenšie nájdené minimum (Global): \nx = %.4f \nF(x) = %.4f\n", xmin, bestmin);
fprintf("==============================================\n");

% Zvýraznenie absolútne najlepšieho bodu v grafe
plot(xmin, bestmin, 'rs', 'MarkerSize', 25, 'LineWidth', 3, 'DisplayName', 'Best Global');
xlabel('x'); ylabel('y = F(x)');
title(['Multi-start Hill Climbing (', num2str(max_iter), ' pokusov)']);

% --- FUNKCIA PRE HĽADANIE ---
function [x_curr, F_curr] = hill_climbing_search(x_curr, F_curr, d)    
    while true
        x_left = x_curr - d;
        x_right = x_curr + d;

        % Ošetrenie hraníc (aby sme nevyskočili z -1000 až 1000)
        if x_left < -1000, F_left = inf; else, F_left = testfn3c(x_left); end
        if x_right > 1000, F_right = inf; else, F_right = testfn3c(x_right); end

        % Porovnanie aktuálneho bodu so susedmi
        [val, idx] = min([F_curr, F_right, F_left]);

        if idx == 2 % Doprava je to lepšie
            x_curr = x_right;
            F_curr = F_right;
        elseif idx == 3 % Doľava je to lepšie
            x_curr = x_left;
            F_curr = F_left;
        else
            % Sme v minime (ani jeden sused nie je lepší)
            plot(x_curr, F_curr, 'bo', 'MarkerFaceColor', 'g'); % Zelený bod pre lokálne minimum
            
            break;
        end

        % Vizualizácia "kráčania"
        plot(x_curr, F_curr, 'bo', 'MarkerSize', 3);
        pause(0.01); % Zrýchlil som pauzu, aby si nečakal večnosť pri 30 iteráciách
    end
end


